### PR TITLE
Add Espresso types and define justification required for L2 batches

### LINKED ISSUE #15: Add `justification` field to `PayloadAttributes`
The HotShot integration will place some additional requirements on each L2 batch, to ensure that it is deterministic. Verifying these requirements requires some additional data as justification, and this data must be included in the `PayloadAttributes` and sent to L1 to ensure that the L2 batches can be derived entirely from state on L1.

The justification consists of an additional field `justification L2BatchJustification`, where:
```go
type L2BatchJustification struct {
  prevBatchLastBlock HotShotBlockHeader
  firstBlock HotShotBlockHeader
  // blockInclusion MerkleTree
  payloadJustification *L2BatchPayloadJustification
}

type L2BatchPayloadJustification {
  lastBlock HotShotBLockHeader
  nextBatchFirstBlock HotShotBlockHeader
  nmtProofs []NmtProof
}

type HotShotBlockHeader struct {
  L1Block HotShotL1BlockInfo
  transactions NmtRoot
}

type HotShotL1BlockInfo struct {
  number big.Int
  timestamp uint64
}
```

* `prevBatchLastBlock` is the header of the last HotShot block of the previous L2 batch, in order to check that the current L2 batch is complete (the following HotShot block is indeed the first of the batch)
* `firstBlock` is the header of the first HotShot block to be included in this batch. It is used to derive the L1 origin
* `blockInclusion` proves the inclusion of each HotShot block commitment in the Merkle tree maintained by the sequencer contract. It is not necessary for now because the sequencer contract directly stores all block commitments
* `payloadJustification` applies when whenever the L2 batch payload is not empty due to an invalid HotShot block. It includes the last block header and the following block header, so that we can check that the range of included HotShot blocks is correct, and an NMT proof relative to each HotShot block commitment in the range. Together, these NMT proofs should justify the `Transactions` field of the `PayloadAttributes`.

For now, the types `MerkleTree`, `NmtRoot`, and  `NmtProof` can be uninterpreted blobs, strings, or JSON objects (whatever is easier) since verification of Merkle proofs will be mocked at first.

### PR DESCRIPTION
**Pull request by jbearer**
Define the HotShot types that the OP derivation pipeline will need to work with -- block headers and related proof types -- and define the extra information that must be included in a batch to show that that batch has been correctly constructed from a range of HotShot blocks.

The L2BatchJustification type is added to the OP chain in two places:
1. It is added to BatchInfo, so that it can be serialized and sent to the batch inbox on L1. This preserves the invariant that the entire L2 chain can be derived from information present on L1.
2. It is added to the calldata of the system-generated L1 info deposit transaction. This is really just a convenient place to put it so that it gets serialized into the L2 block without changing the L2 block type (which is currently identical to the L1 block type). Doing this allows us to reconstruct batches _with justification_ from blocks, which is important because the batches submitted by the batcher are derived from blocks produced by the execution engine. By adding the justification to both the batch and the block, we ensure that it will be correctly serialized and sent to L1 by the batcher, then deserialized and recovered from L1 by the derivation pipeline, without requiring further changes to either (except the business logic of populating the justification in the batcher and validating the batch constraints in the batch queue).

This change attempts to preserve support for the old BatchV1 type without this justification, so that
* We can still run the node without Espresso
* The node will be able to replay a chain which started before Espresso and then added support for Espresso some time later

I had to change many, many tests, but on the bright side one of those tests caught an actual bug -- the `nil` edge case in `L1BlockInfo.MarshalBinary()`.

The next step is to implement the basic operations that we will need on the HotShot types. Since NMT proof validation will be mocked for a while, the main operation we need is the ability to compute a HotShot block commitment from a `Header`, so that we can authenticate the headers in the justification against certified block commitments from the sequencer contract.

This change closes #15.

### CODE REVIEW

File: op-node/eth/types.go [hunk 1]
```diff
@@ -236,6 +238,33 @@ type PayloadAttributes struct {
...
+	FirstBlock espresso.Header `json:"firstBlock"`
+
+	// The block number of `FirstBlock`. This allows us to authenticate the `FirstBlock` header by
+	// computing the commitment of `FirstBlock` and checking it against the commitment at position
+	// `FirstBlockNumber` in a certified sequence (e.g. the sequencer contract). We can also
+	// authenticate `PrevBatchLastBlock` in a similar way, since it is required to be the block
+	// immediately before `FirstBlock`.
+	FirstBlockNumber uint64 `json:"firstBlockNumber"`
+
+	Payload *L2BatchPayloadJustification `json:"payload,omitempty",rlp:"nil"`
```
Comment (nomaxg) [inline op-node/eth/types.go:252]: Will the serialization panic without this rlp:"nil" tag?

Comment (nomaxg) [inline op-node/eth/types.go:252]: Why have this sub-struct? Might be cleaner to keep things flat if possible

File: op-node/eth/types.go [hunk 2]
```diff
@@ -236,6 +238,33 @@ type PayloadAttributes struct {
 	GasLimit *Uint64Quantity `json:"gasLimit,omitempty"`
 }
 
+type L2BatchJustification struct {
+	PrevBatchLastBlock espresso.Header `json:"prevBatchLastBlock"`
+	FirstBlock espresso.Header `json:"firstBlock"`
+
+	// The block number of `FirstBlock`. This allows us to authenticate the `FirstBlock` header by
```
Comment (nomaxg) [inline op-node/eth/types.go:245]: Can we store the block number in the header? Or is this not possible right now because `next_block` is stateless?

File: op-node/eth/types.go (hunk 1 above)
Comment (jbearer) [reply op-node/eth/types.go:252]: This affects deserialization. It makes it so that an empty RLP list deserializes as `nil`. Without this tag, the `rlp` library requires the data to represent a non-nil object or else deserialization fails. This field is specifically intended to be optional so we need to allow `nil`

Comment (jbearer) [reply op-node/eth/types.go:252]: I should add a comment. It's because this sub-struct is optional, it's omitted in the various exception cases where we skip HotShot blocks that don't satisfy the constraints

File: op-node/eth/types.go (hunk 2 above)
Comment (jbearer) [reply op-node/eth/types.go:245]: Exactly, `next_block` doesn't have access to the block number that's being created. It's ok though because this is authenticated by looking up the block at this position in the contract, and not via the block commitment scheme

**Review by nomaxg — APPROVED** [review]
LGTM, just a few comments

File: op-node/eth/types.go (hunk 1 above)
Comment (nomaxg) [reply op-node/eth/types.go:252]: ok sounds good
